plugin.py

```diff
@@ -46,77 +45,83 @@ def prepare(self):
+
     # auto route by invoking FreeRouting.jar
     def RunRouter(self):
-        command = [self.java_path, "-jar", self.module_path, "-de", self.module_input, "-s"]
-        try:
-            process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-            output, error = process.communicate()
-            if process.returncode == 0:
+        command = self.module_command
+        dialog = wx_make_cancel("""
+        Complete or Terminate FreeRouting:
+        * to complete, close Java window
+        * to terminate, press Terminate here
+        """)
+        invoker = ProcessThread(command, dialog)
+        dialog.Show()  # dialog first
+        invoker.start()  # run java process
+        result = dialog.ShowModal()  # block pcbnew here
+        if result == wx.ID_OK:  # return via terminate button
+            invoker.terminate()
+            return False
+        elif result == wx.ID_CANCEL:  # return via dialog.Destroy()
+            if invoker.has_ok():
                 return True
-            else:  
-                show_error("""
-                Java program failure:
-                exit code: %d
-                --- std output ---
-                %s
-                --- std error  ---
-                %s
-                """ % (process.returncode, output, error))
-                return False
-        except Exception as error:
-                show_error("""
-                Java process failure: 
-                %s""" % str(error))
+            else:
+                invoker.show_error()
                 return False
+        else:
+            return False  # should not happen
 
     # import generated board.ses file into pcbnew
     def RunImport(self):
         ok = pcbnew.ImportSpecctraSES(self.module_output)
         if ok and os.path.isfile(self.module_output):
             return True
         else:
-            show_error("""
+            wx_show_error("""
             Failed to invoke:
             * pcbnew.ImportSpecctraSES    
             """)
             return False
         
     # invoke chain of dependent methods
     def RunSteps(self):
-        return self.RunExport() and self.RunRouter() and self.RunImport()
+        self.prepare()
+        if not self.RunExport() :
+            return 
+        if not self.RunRouter() :
+            return
+        wx_safe_invoke(self.RunImport)
 
     # kicad plugin action entry
     def Run(self):
-        if not has_pcbnew_api():
-            show_error("""
+        if has_pcbnew_api():
+            self.RunSteps()
+        else:
+            wx_show_error("""
             Missing required python API:
             * pcbnew.ExportSpecctraDSN
             * pcbnew.ImportSpecctraSES
             ---
             Try development nightly build:
             * http://kicad-pcb.org/download/
             """)
-            return
-        self.prepare()
-        safe_invoke(self.RunSteps)
 
 
 # provision gui-thread-safe execution context
```

Comment:
You don't need this anymore and it actually prevents KiCad process from exiting.

Reply:
I do not follow.
1. wx.CallAfter required for import board.ses
2. wx.App instance is required by wx.CallAfter

Reply:
You don't even reference theApp anywhere. And when you run in KiCad you already have a wxApp (it's c++ counterpart) initialized because kicad itself is a wxwidgets app. So wx.CallAfter works just fine.

Reply:
Also on linux because of extra wxApp created here the kicad process exits cleanly and when you restart it you get "kicad is already running, are you sure?".

Reply:
yes, see #1